### DISCUSSION TITLE
How to edit the final html output (after he's been rendered)

### DISCUSSION THREAD

**Discussion by maximelebreton** [nuxt/nuxt#20059] [General]
Hello there!

I have a question.
I would like to generate an **aside** menu, based on all the **h2** in my html page.

So here is the **original output** of my page:
```html
<div class="page__body">
  <a id="title-1"></a>
  <h2>Title 1</h2>
  <p>Lorem Ipsum</p>

  <a id="title-2"></a>
  <h2>Title 2</>
  <p>Ipsum Lorem</p>
</div>
<div class="page__aside"></div>
```  

And I would like to generate this markup inside .page__aside:
```html
<div class="page__aside">
<ul>
  <li><a href="#title-1">Title 1</a></li>
  <li><a href="#title-2">Title 2</a></li>
</ul>
</div>
``` 

How can I achieve this?

I start to play with the `render:html` method:
```ts
import jsdom from "jsdom";

export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook("render:html", async (html, { event }) => {
    const { JSDOM } = jsdom;
    const { document } = new JSDOM(html).window;

    const pageBodyElement = document.querySelector(".page__body");
    const asideUlElement = document.querySelector(
      ".page__aside ul"
    );

    const h2Elements = pageBodyElement?.querySelectorAll("h2");

    const liElements = [...h2Elements!].map((element) => {
      return `<li><a href="${element.previousElementSibling?.id}">${element.innerText}</a></li>`;
    });
    const ulHtml = `
      ${liElements.map((html) => html)}
    `;
    asideUlElement?.append(ulHtml);
  });
});
```

But after put a console.log inside, it seems that the final html content is **still not rendered** in the `render:html` lifecycle step.

So wich hook can I use to get the final html ouput, and customize it?

Thanks!

**Comment by warflash** [comment]
Without knowing the usecase I'd say this seems like a very dangerous and hard to maintain approach. Here is how vitepress is generating its table of contents using a composable, maybe that serves as inspiration: https://github.com/vuejs/vitepress/blob/399aeea7d0c2b9f8274c0c4b93ec2bfdfe98a28f/src/client/theme-default/composables/outline.ts
